src/augmented_dickey_fuller.jl

```diff
@@ -0,0 +1,219 @@
+
+immutable ADFTest <: HypothesisTest
+    n               ::Int               # number of observations
+    deterministic   ::AbstractString    # deterministic terms included in regression
+    lag             ::Int               # number of lags in test statistic
+    stat            ::Float64           # test statistic
+    coef            ::Float64           # coefficient on lagged (non-differenced) variable
+    cv              ::Vector{Float64}   # critical values
+end
+
+"""
+    ADFTest(y, deterministic, lag)
+
+Compute the augmented Dickey-Fuller unit root test.
+
+`y` is the time series to be tested, `deterministic` determines the deterministic terms
+(options: `none`, `constant`, `trend`, `squared_trend`) and `lag` the number of lagged
+first-differences included in the test regression, respectively.
+
+Critical values and asymptotic p-values are computed based on response surface regressions
+following MacKinnon (2010) and MacKinnon (1994), respectively. These may differ slightly
+from those reported in other regression packages as different algorithms might be used.
+
+References
+
+  * James G. MacKinnon, 2010, "Critical values for cointegration tests,"
+    QED Working Paper No. 1227, 2010, [link](http://ideas.repec.org/p/qed/wpaper/1227.html).
+  * James G. MacKinnon, 1994, "Approximate Asymptotic Distribution Functions for
+    Unit-Root and Cointegration Tests", Journal of Business & Economic Statistics,
+    Vol. 12, No. 2, pp. 167-176, [link](http://www.jstor.org/stable/1391481).
+
+External links
+
+  * [Augmented Dickey-Fuller test on Wikipedia](https://en.wikipedia.org/wiki/Augmented_Dickey–Fuller_test)
+"""
+function ADFTest{T<:Real}(y::AbstractVector{T}, deterministic::AbstractString, lag::Int)
+
+    nobs = length(y)
+    Δy = diff(y)
+
+    if deterministic == "none"
+        Δyt   = Δy[lag+1:end]
+        xt    = (y[lag+1:end-1]')' # to make sure that xt is always two dimensional Array
+                                   # needed in line 94 for case with no lags
+        dfexo = 0
+    elseif deterministic == "constant"
+        Δyt   = Δy[lag+1:end]
+        xt    = [ y[lag+1:end-1] ones(nobs-lag-1, 1) ]
+        dfexo = 1
+    elseif deterministic == "trend"
+        Δyt   = Δy[lag+1:end]
+        xt    = [ y[lag+1:end-1] ones(nobs-lag-1, 1) (1:nobs-lag-1) ]
+        dfexo = 2
+    elseif deterministic == "squared_trend"
+        Δyt   = Δy[lag+1:end]
+        xt    = [ y[lag+1:end-1] ones(nobs-lag-1, 1) (1:nobs-lag-1) [
+                  i^2 for i = 1:nobs-lag-1] ]
+        dfexo   = 3
+    else
+        throw(ArgumentError("deterministic = $(deterministic) is invalid"))
+    end
+
+    if lag > 0
+        Δylag = hcat([Δy[lag-i+1:end-i] for i=1:lag]...)
+        xt = [ xt Δylag ]
+    end
+
+    adf_coef = xt \ Δyt
+    adf_res  = Δyt - xt * adf_coef
+    n        = length(adf_res)
+    k        = lag + 1 + dfexo
+    sigma2   = dot(adf_res, adf_res) / (n - k)
+    cov_coef = sigma2 * inv(xt' * xt)
```

Comment:
You only use element `1,1` so you can avoid a temporary array by postponing the multiplication of `sigma2` to next line.